AUTOSCALE 267 - Create KedaController upon keda-olm-operator Installation

fishereskew opened the pull request:
Draft PR - Currently installs default keda controller (and all keda components) upon operator installation. Tested locally using minikube and crc. I have not written any tests yet, I've only ensured the basic use case works. I have some TODO comments for questions and uncertainties in internal/controller/keda/kedacontroller_controller.go.

joelsmith reviewed (changes requested):
This is looking really great! Only a couple of small issues and then I think it will be ready to merge

maxcao13 commented on internal/controller/keda/kedacontroller_controller.go:
```diff
@@ -115,12 +116,102 @@ func (r *KedaControllerReconciler) SetupWithManager(mgr ctrl.Manager, kedaContro
+
+	// Fetch operator namespace
+	kedaNamespace := &corev1.Namespace{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: r.resourceNamespace}, kedaNamespace)
```
Oops never mind, I took a look again and I misinterpreted what the line was doing. Sorry, don't worry about that

jkyros reviewed (commented):
Oh no, it's John! :stuck_out_tongue: 

Fisher this is great, thanks so much for your effort here!  Kubernetes controllers are weird. If you want more context, or you want to talk through any of this synchronously let me know.

maxcao13 commented on config/manager/kustomization.yaml:
```diff
@@ -1,2 +1,8 @@
+kind: Kustomization
+images:
+- name: ghcr.io/kedacore/keda-olm-operator
+  newName: quay.io/rh-ee-feskew/keda-olm-operator/keda-olm-operator
```
Yeah this gets changed every time you do a make publish or something, so just something to be wary of when you commit. We don't want these types of changes to be in there.

fishereskew commented on internal/controller/keda/kedacontroller_controller.go:
```diff
@@ -115,12 +116,106 @@ func (r *KedaControllerReconciler) SetupWithManager(mgr ctrl.Manager, kedaContro
+		logger.Error(err, "Failed to get Keda namespace to check annotation.")
+		return
+	}
+	kedaNamespaceCopy := kedaNamespace.DeepCopy()
```
I had it where it is there because it was right after getting the object and doing the error checking. I'll move it into the if statement though because it doesn't really matter where it is and saving memory is important! Thanks!

fishereskew commented on internal/controller/keda/kedacontroller_controller.go:
```diff
@@ -115,12 +116,102 @@ func (r *KedaControllerReconciler) SetupWithManager(mgr ctrl.Manager, kedaContro
+	// Fetch operator namespace
+	kedaNamespace := &corev1.Namespace{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: r.resourceNamespace}, kedaNamespace)
+	if ctxerr := ctx.Err(); ctxerr != nil {
```
You are correct. I checked to make sure and the context timeout propagates back and sets err to contain the context timeout error, so all of the extra context checks can be removed. Thanks!

fishereskew commented:
Updated PR:

- Removed unnecessary context error checks
- Added more info for patch error message
- Changed kedaNamespace to not be modified directly, but to instead modify a copy
- Fixed Patch() so that the copy is patched onto the actual object. Also switched to StrageticMergeFrom() instead of MergeFrom()

maxcao13 reviewed (commented):
Generally looks good to me :)

For fun, I ran the cma tests with this new setup configuration and they passed meaning the operator changes didn't somehow mess and of the operands up:
```
E2E_INSTALL_KEDA=false make e2e-test-openshift-setup
make e2e-test-openshift
```

maxcao13 commented on config/manager/kustomization.yaml:
```diff
@@ -1,2 +1,2 @@
 resources:
-- manager.yaml
+- manager.yaml
```
I think you removed the newline at the end of the file.

If not, then you can just remove all changes you've ever done to that file by (which we probably should be doing anyways to make the git history for that file not ugly)
```
git rebase -i kedacore/main
```
editing your commits like
```
edit 7de807a3 # controller created upon operator installation
edit e99027e9 # fix: context error checking and added test cases
```
and "resetting" any changes you've done by checking out the file from main:
```
git checkout kedacore/main -- config/manager/
```
then resolving the rebase conflict that will happen by selecting "current" changes.

then git diff should come out clean after:

```
git diff kedacore/main -- config/manager/kustomization.yaml
```

joelsmith commented:
It looks like the CI static checks found a couple of nits:

```
internal/controller/keda/kedacontroller_controller_test.go:136:19: Error return value of `k8sClient.Get` is not checked (errcheck)
					k8sClient.Get(ctx, types.NamespacedName{Name: namespace}, kedaNamespace)
					             ^
internal/controller/keda/kedacontroller_controller.go:186:1: unnecessary trailing newline (whitespace)
```

Maybe you could catch any error and if one is returned, you could return it as a string so that if the test eventually times out before succeeding, we'll see the error when it complains that "my fake sample error here" != "true"

joelsmith reviewed (approved):
Awesome, let's merge this thing!
